Read progress tracking phase 1

DrzikFilip opened the pull request.

DrzikFilip committed:
Read progress tracking phase 1

zoldia committed:
add mail type category show title column migration

remp/crm#660

miroc committed:
Fix generation of images in DennikN content

DrzikFilip committed:
updates based on testing

zoldia committed:
remove redundant get locked html methods from generators

DrzikFilip committed:
send request fix

DrzikFilip committed:
update custom event

miroc committed:
Fix initialization of Smtp Mailer

miroc committed:
Merge branch 'master' of https://gitlab.com/remp/remp

* 'master' of https://gitlab.com/remp/remp:
  remove redundant get locked html methods from generators

miroc committed:
Fix rendering images for Dennikn content

rootpd committed:
Adding ProcessPageviewSessions to Beam's console Kernel

This command was unintentionally omitted from Kernel and the default
installation doesn't have any session/device data because of that.
Even though it's still possible to run the command manually, this
shouldn't be visible to the user/admin and it should be consistent
with the rest of command suite.

rootpd commented on Beam/resources/assets/js/remplib.js:
```diff
@@ -98,6 +106,15 @@ remplib = typeof(remplib) === 'undefined' ? {} : remplib;
                 if (typeof config.tracker.article.locked !== 'undefined') {
                     this.article.locked = config.tracker.article.locked;
                 }
+                if (typeof config.articleElementFn() !== 'undefined') {
```
* Ked je nieco funkcia / callback, chceme to zavolat co "najneskor".
* Ak ta funkcia nie je definovana, cele to tu spadne a nic dalsie sa nevykona.

rootpd commented on Beam/resources/assets/js/remplib.js:
```diff
@@ -98,6 +106,15 @@ remplib = typeof(remplib) === 'undefined' ? {} : remplib;
+                if (typeof config.articleElementFn() !== 'undefined') {
+                    this.trackedArticle = config.articleElementFn()
+                }
+                if (typeof config.tracker.readingProgress.interval !== 'undefined') {
```
Treba este skontrolovat, ci je vobec `readingProgress` zadefinovany. Vsetky povinne parametre sa kontroluju hned na zaciatku inicializacie a vyhadzuju chybu v pripade, ze nieco chyba.

`readingProgress` ako taky ale nemusi byt povinny. Vtedy by si mal pouzit defaulty, ktore som spominal v paperi:

* Ak nie je `readingProgress` alebo nie je `readingProgress.enabled`, tak by mal byt niekde ulozeny flag, ze reading progress netrackujeme
* Ak nie je `readingProgress` alebo nie je `readingProgress.interval`, tak by mal byt niekde stale ulozeny defaultny interval "5 sekund"

rootpd replied:
Teraz este dodatocne pozeram, ze nizsie tu podmienku mas, toto si asi omylom dal do zlej `if` vetvy patriacej clanku.

rootpd commented on Beam/resources/assets/js/remplib.js:
```diff
@@ -572,6 +600,77 @@ remplib = typeof(remplib) === 'undefined' ? {} : remplib;
+                page.article = Math.min( 1, Math.max( 0, ( currentPosition - article.offsetTop) / article.offsetHeight ));
+            }
+
+            var event = document.createEvent('CustomEvent'); // old legacy way of defining custom events, because of IE support
```
Toto nie je nutne. V zdielanej casti `remplib.js` je na toto nachystany uz polyfill. Mozes kludne zavolat 

```
 let event = new CustomEvent("foo", {
    bar: "baz",
});
window.dispatchEvent(event);
```

Podobne je to pouzite aj par riadkov vyssie.

rootpd commented:
Pri testovani dostavam z Tracker API HTTP 400 Bad request.

```js
{
    "id": "l3EQRT3L",
    "code": "bad_request",
    "status": 400,
    "detail": "[hWleeeNI] 400 invalid_request: attribute \"system\" of response is missing and required; attribute \"user\" of response is missing and required, attribute: user, parent: response"
}
```

Chybaju tam povinne parametre. Pred zavolanim treba este do parametrov pridat user/system data zavolanim `params = this.addSystemUserParams(params);`.

Testnite to prosim aj voci tomu Tracker API (je sucastou Dockra).

rootpd reviewed (commented):
Pri testovani dostavam z Tracker API HTTP 400 Bad request.

```js
{
    "id": "l3EQRT3L",
    "code": "bad_request",
    "status": 400,
    "detail": "[hWleeeNI] 400 invalid_request: attribute \"system\" of response is missing and required; attribute \"user\" of response is missing and required, attribute: user, parent: response"
}
```

Chybaju tam povinne parametre. Pred zavolanim treba este do parametrov pridat user/system data zavolanim `params = this.addSystemUserParams(params);`.

Testnite to prosim aj voci tomu Tracker API (je sucastou Dockra).

miroc committed:
Refactoring

miroc committed:
Show conversion events in article loads graph

miroc committed:
Event options on article loads graph

miroc committed:
Refactoring

miroc committed:
Private constant

rootpd committed:
Bugfixing mail type stats view when no stats are present yet

miroc committed:
Check for undefined

miroc committed:
Merge branch '401-conversions_shown_in_page_loads_graph' into 'master'

do grafu zivotneho cyklu clanku pridat ciaru pre kazdu konverziu

See merge request remp/remp!283

miroc committed:
Show article conversions in graph by default

miroc committed:
Compute a/b test flags for top articles in dashboard

miroc committed:
Graphics for A/B test flags on dashboard

miroc committed:
Do not set a/b test flags based on whole articles history

miroc committed:
Check for ab test flags in last 5 minutes only

DrzikFilip committed:
fixes based on comments

DrzikFilip commented on Beam/resources/assets/js/remplib.js:
```diff
@@ -657,7 +662,7 @@ remplib = typeof(remplib) === 'undefined' ? {} : remplib;
 
-            const params = {
+            var params = {
                 "action": "progress",
```
@rootpd Pri testovani mi backend vratil chybu ze action moze byt len "load" alebo "timespent". Je to vec ktoru my riesit nemusime ?

DrzikFilip commented on Beam/resources/assets/js/remplib.js:
```diff
@@ -657,7 +662,7 @@ remplib = typeof(remplib) === 'undefined' ? {} : remplib;
 
-            const params = {
+            var params = {
                 "action": "progress",
```
@rootpd Pri testovani mi backend vratil chybu ze action moze byt len "load" alebo "timespent". Je to vec ktoru my riesit nemusime ?

rootpd replied:
Mozem vidiet cely payload? Mas urcite aktualnu verziu trackeru? https://github.com/remp2020/remp/releases/tag/0.5.2

```
➜  remp git:(master) ✗ curl -v -X POST \
  http://tracker.beam.remp.press/track/pageview \
  -H 'Content-Type: application/json' \
  -H 'Origin: http://campaign.remp.press' \
  -d '{
    "article": {
        "id": "9900222",
        "author_id": "autoresku blesku 222",
        "category": "kategoria 222",
        "locked": true,
        "tags": ["tag1", "tag2", "tag3"],
        "variants": {}
    },
    "action": "progress",
    "progress": {
        "page_ratio": 0.203,
        "article_ratio": 0.95
    },
    "system": {
        "property_token": "1a8feb16-3e30-4f9b-bf74-20037ea8505a",
        "time": "2019-01-25T15:21:27.293Z"
    },
    "user": {
        "browser_id": "0fa7e641-7573-4ccd-90da-c7b902b08755",
        "subscriber": true,
        "referer": "",
        "adblock": true,
        "window_height": 1243,
        "window_width": 1603,
        "cookies": true,
        "websockets": true,
        "source": {
            "ref": "direct"
        },
        "remp_session_id": "2d06f9c3-2e73-4c53-9407-a5958876ef7b",
        "remp_pageview_id": "34eeae3b-0223-422e-9175-f8be71a3c6ac"
    }
}'
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to tracker.beam.remp.press (127.0.0.1) port 80 (#0)
> POST /track/pageview HTTP/1.1
> Host: tracker.beam.remp.press
> User-Agent: curl/7.61.1
> Accept: */*
> Content-Type: application/json
> Origin: http://campaign.remp.press
> Content-Length: 930
> 
* upload completely sent off: 930 out of 930 bytes
< HTTP/1.1 202 Accepted
< Server: nginx/1.14.2
< Date: Mon, 13 May 2019 07:36:28 GMT
< Content-Length: 0
< Connection: keep-alive
< Access-Control-Allow-Credentials: false
< Access-Control-Allow-Origin: http://campaign.remp.press
< 
* Connection #0 to host tracker.beam.remp.press left intact
```

DrzikFilip replied:
Myslim, ze mame najaktualnejsiu verziu. Dnes sme to skusali aj s tou najaktualnejsou. Tu je teda cely payload. 

`curl 'http://tracker.beam.remp.press:8000/track/pageview' -H 'Referer: http://beam.remp.press:8000/test.html' -H 'Origin: http://beam.remp.press:8000' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36' -H 'Content-Type: application/json' --data-binary '{"action":"progress","progress":{"page_ratio":0.48602673147023084,"unload":false,"article_ratio":0.2549728752260398},"system":{"property_token":"8f406aaa-4e70-35cc-a054-903acd1ca587","time":"2019-05-14T14:48:46.165Z"},"user":{"id":"1","browser_id":"d6f5c542-fdd7-4aff-852b-012509a0b4a1","subscriber":true,"url":"http://beam.remp.press:8000/test.html","referer":"","user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36","adblock":false,"window_height":856,"window_width":1409,"cookies":true,"websockets":true,"source":{},"remp_session_id":"c5db358b-e51f-4ee1-9f5a-05dcf647629b","remp_pageview_id":"e07a962a-70e1-4075-9d47-ee887759e5bc"}}' --compressed`


Vracia to toto: 
`{"id":"PP5mwimP","code":"bad_request","status":400,"detail":"[Eif05XET] 400 invalid_request: value of response.action must be one of \"load\", \"timespent\" but got value \"progress\", attribute: response.action, value: progress, expected: \"load\", \"timespent\""}
`

rootpd replied:
Rovnaky payload vracia u mna 202 accepted (vymenil som len `property_token`, ten mame logicky iny). Ak si teda v dockri, spusti v roote projektu:

```
make docker-build
docker-compose stop beam_tracker
docker-compose build beam_tracker
docker-compose up -d beam_tracker
```

Ak si mimo dockeru, stiahni si aktualnu binarku tu: https://github.com/remp2020/remp/releases/tag/0.5.3

miroc committed:
Fix for loading conversions with wrong timezone

rootpd committed:
Fixing invalid name of index in telegraf.conf and wrong snippet in Mailer's README.

rootpd committed:
Adding alias support for elastic mapping caching in Segments API

If alias is used for index name on elastic server, Segments API couldn't
get mapping information for that segment. If such case occurs, db_elastic.go
tries to get available aliases for requested index as a fallback
and use latest index (alphabetically) to get mapping information.

rootpd committed:
Bugfixing greedy conversion stats fetching

rootpd reviewed (commented):
Ked doscrollujem uplne na koniec stranky, libka zatrackuje percenta z horneho okraja obrazovky. S tymto by si trackovacia kniznica mala poradit a ked clovek doscrolluje uplne dole, mala by zatrackovat `1.000`.

Takyto payload som dostal ked som doscrolloval dole na testovacom clanku, ktory je vysoky takmer dve obrazovky.

```json5
progress: {
    page_ratio: 0.42284325637910086,
    unload: false,
    article_ratio: 0.1609403254972875
}
```

Ocakaval by som v oboch pripadoch `ratio = 1`

matejminar committed:
Merge branch 'master' of github.com:BRACKETS-by-TRIAD/remp

matejminar committed:
Merge branch 'RPD-phase-1' of github.com:BRACKETS-by-TRIAD/remp into RPD-phase-1

matejminar committed:
scroll progress refactor

matejminar committed:
progress computing throttled, only tracking max progress (not going back), removed fast scroll check

matejminar committed:
removed unnecessary code (legacy fast scrolling check)

matejminar committed:
unload handler fixed to not fire at load
